### PR TITLE
:sparkles: [AnalysisWizard]  Language discovery changes 

### PR DESCRIPTION
**Pull request by ibolton336**
Resolves https://github.com/konveyor/tackle2-ui/issues/1950

UI Tests PR: 1136

Needs: https://github.com/konveyor/tackle-ui-tests/pull/1136

Includes: 
- Provide a "(Show All)" option to display all targets regardless of language tags. Adds a new component to handle this  new menu type. 
- Populate the options list from the Provider field of the fetched targets.
- Initially select the languages based on the tags, of tag category "Language", across the applications selected for analysis.
- Update TS type for Target to reflect the provider type changing to a string[].
- Covers the hub change to the task model described here:
>The application Analysis fields used to correlated to task by addon. This needs to be updated to correlate by kind == "analyzer" (when not blank) else fallback on addon == "analyzer". The fallback is needed to correlate tasks created in previous releases.
- Each target card should always display a label with it's provider. This will allow cards that belong to different providers to be differentiated when more than one language is selected.

Screenshot 2024-06-12 at 3 37 47 PM

### CODE REVIEW

File: client/src/app/components/SimpleSelectCheckbox.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    ]
+  );
+
+  React.useEffect(() => {
+    setSelectedItems(value || []);
+  }, [value]);
+
+  React.useEffect(() => {
+    const updatedOptions = [
+      { value: "select-all", label: "Select All", children: "Select All" },
```
Comment (rszwajko) [inline client/src/app/components/SimpleSelectCheckbox.tsx:52]: IMHO  "select" is not well defined here.  Above, we ask user "select one or more target options..." (target === card).  User may think that by selecting a provider/language they select a target. 
Perhaps we should switch to filtering-related wording i.e. "show all" , "Filter by provider" ....

File: client/src/app/pages/applications/analysis-wizard/set-targets.tsx [hunk 2]
```diff
@@ -15,24 +16,53 @@ import { TargetCard } from "@app/components/target-card/target-card";
...
+        .map((tag) => {
+          return {
+            category: findCategoryForTag(tag.id),
+            tag,
+          };
+        })
+        .filter((tagWithCat) => tagWithCat?.category?.name === "Language")
+        .map((tagWithCat) => tagWithCat.tag.name)
+    )
+  ).filter((name) => name !== undefined);
```
Comment (rszwajko) [inline client/src/app/pages/applications/analysis-wizard/set-targets.tsx:62]: `filter(Boolean)` ?

File: client/src/app/queries/tasks.ts [hunk 3]
```diff
@@ -27,7 +28,16 @@ export const useFetchTasks = (
     select: (allTasks) => {
       const uniqSorted = allTasks
         .filter((task) =>
-          filters?.addon ? filters.addon === task.addon : true
+          // If there are any tasks with the addon field, we will still need to consider those older
```
Comment (rszwajko) [inline client/src/app/queries/tasks.ts:31]: I would move it to a separate PR. First, it seems not directly lined with the enhancement. Second, this is really should be re-written with server side filtering (https://github.com/konveyor/tackle2-ui/pull/1956)

Comment (ibolton336) [reply client/src/app/queries/tasks.ts]: Moved to #1959

File: client/src/app/components/SimpleSelectCheckbox.tsx (hunk 1 above)
Comment (ibolton336) [reply client/src/app/components/SimpleSelectCheckbox.tsx]: +1 Changed to show verbiage.

**Comment by sjd78** [comment]
e2e ui testing is failing: https://github.com/konveyor/tackle2-ui/actions/runs/9504942845/job/26198861030?pr=1951#step:10:1822

**Review by sjd78 — COMMENTED** [review]
First look makes sense, but there is an inconsistency with the Target.provider field.

**Comment by ibolton336** [comment]
> First look makes sense, but there is an inconsistency with the Target.provider field.

Fixed.

File: client/src/app/components/SimpleSelectCheckbox.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  options,
+  value,
+  placeholderText = "Select...",
+  id,
+  toggleId,
+  toggleAriaLabel,
+  width,
+}) => {
+  const [isOpen, setIsOpen] = React.useState(false);
+  const [selectedItems, setSelectedItems] = React.useState<string[]>([]);
```
Comment (rszwajko) [inline client/src/app/components/SimpleSelectCheckbox.tsx:38]: The selection state and options are maintained already by the parent (`provider` and  `languageOptions`).
Can we rely only on them and remove local state here?

File: client/src/app/pages/applications/analysis-wizard/set-targets.tsx [hunk 5]
```diff
@@ -15,24 +16,53 @@ import { TargetCard } from "@app/components/target-card/target-card";
...
+            category: findCategoryForTag(tag.id),
+            tag,
+          };
+        })
+        .filter((tagWithCat) => tagWithCat?.category?.name === "Language")
+        .map((tagWithCat) => tagWithCat.tag.name)
+    )
+  ).filter(Boolean);
+
+  const [provider, setProvider] = useState(initialProviders);
```
Comment (rszwajko) [inline client/src/app/pages/applications/analysis-wizard/set-targets.tsx:64]: renaming the field to `providers` would improve readability.

File: client/src/app/components/SimpleSelectCheckbox.tsx (hunk 4 above)
Comment (ibolton336) [reply client/src/app/components/SimpleSelectCheckbox.tsx:38]: I think it makes sense to have internal component level state here since we are manipulating it based on the showAll checkbox value.

**Comment by ibolton336** [comment]
> Looks good! I've found one bug during smoke tests:
> 
> 1. filter by "Java" and select some cards
> 2. go back in the wizard
> 3. enter the target selection step again: filter is missing, cards cannot be de-selected

Looks like this bug shows up as a result of introducing checkboxes for unselected cards:

```
    isSelectable={!!cardSelected}

```

I'd like to fix this in another issue if possible. Need to investigate the PF component. Reverted the change for now & https://github.com/konveyor/tackle2-ui/issues/1252 will remain open.

**Comment by rszwajko** [comment]
@ibolton336 
Integration tests are failing at selecting the card:
```
AssertionError: Timed out retrying after 8000ms: Expected to find element: `input[type="checkbox"]`, but never found it. Queried from:
              > cy.get(#target-card-Containerization)
```

**Comment by ibolton336** [comment]
> @ibolton336 Integration tests are failing at selecting the card:
> 
> ```
> AssertionError: Timed out retrying after 8000ms: Expected to find element: `input[type="checkbox"]`, but never found it. Queried from:
>               > cy.get(#target-card-Containerization)
> ```
Fixed.

**Review by rszwajko — APPROVED** [review]
